In [4]:
import pandas as pd

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Stage1. 데이터 불러오기

### Step 1. 데이터 불러오기

In [5]:
# sample_submission.csv 파일을 submission이라는 변수에 할당
submission = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/DAT/4th_team/sample_submission.csv')

train = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/DAT/4th_team/train.csv')
test = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/DAT/4th_team/test.csv')

### Step 2. 간단하게 데이터 살펴보기

#### 1. head()/ tail()

In [6]:
submission.head()

index  0  1  2
0  26457  0  0  0
1  26458  0  0  0
2  26459  0  0  0
3  26460  0  0  0
4  26461  0  0  0

* 모든 데이터가 0으로 이루어져 있기 때문에, 각 `index`에 따른 고객의 신용도를 예측하는 문제임을 알 수 있다
* 따라서, test고객 데이터의 신용도 예측 결과에 대한 확률로 `index`의 값을 모두 채우면 된다.

`index`: 고객번호

`0,1,2`: 신용도 컬럼

In [7]:
# train 데이터의 하위 행 5개를 확인합니다.
train.head()

index gender car reality  child_num  income_total           income_type  \
0      0      F   N       N          0      202500.0  Commercial associate   
1      1      F   N       Y          1      247500.0  Commercial associate   
2      2      M   Y       Y          0      450000.0               Working   
3      3      F   N       Y          0      202500.0  Commercial associate   
4      4      F   Y       Y          0      157500.0         State servant   

                        edu_type     family_type           house_type  \
0               Higher education         Married  Municipal apartment   
1  Secondary / secondary special  Civil marriage    House / apartment   
2               Higher education         Married    House / apartment   
3  Secondary / secondary special         Married    House / apartment   
4               Higher education         Married    House / apartment   

   DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  work_phone  phone  email  \
0      -13899          -4709           1           0      0      0   
1      -11380          -1540           1           0      0      1   
2      -19087          -4434           1           0      1      0   
3      -15088          -2092           1           0      1      0   
4      -15037          -2105           1           0      0      0   

    occyp_type  family_size  begin_month  credit  
0          NaN          2.0         -6.0     1.0  
1     Laborers          3.0         -5.0     1.0  
2     Managers          2.0        -22.0     2.0  
3  Sales staff          2.0        -37.0     0.0  
4     Managers          2.0        -26.0     2.0

#### 2. columns

In [8]:
# train 데이터의 컬럼들을 확인합니다.
train.columns

Index(['index', 'gender', 'car', 'reality', 'child_num', 'income_total',
       'income_type', 'edu_type', 'family_type', 'house_type', 'DAYS_BIRTH',
       'DAYS_EMPLOYED', 'FLAG_MOBIL', 'work_phone', 'phone', 'email',
       'occyp_type', 'family_size', 'begin_month', 'credit'],
      dtype='object')

변수설명
* index
* gender: 성별
* car: 차량 소유 여부
* reality: 부동산 소유 여부
* child_num: 자녀 수
* income_total: 연간 소득
* income_type: 소득 분류
							['Commercial associate', 'Working', 'State servant', 'Pensioner', 'Student']

* edu_type: 교육 수준
							['Higher education' ,'Secondary / secondary special', 'Incomplete higher', 'Lower secondary', 'Academic degree']

* family_type: 결혼 여부
							['Married', 'Civil marriage', 'Separated', 'Single / not married', 'Widow']

* house_type: 생활 방식
							['Municipal apartment', 'House / apartment', 'With parents',

							 'Co-op apartment', 'Rented apartment', 'Office apartment']

* DAYS_BIRTH: 출생일
							데이터 수집 당시 (0)부터 역으로 셈, 즉, -1은 데이터 수집일 하루 전에 태어났음을 의미

* DAYS_EMPLOYED: 업무 시작일
							데이터 수집 당시 (0)부터 역으로 셈, 즉, -1은 데이터 수집일 하루 전부터 일을 시작함을 의미

							양수 값은 고용되지 않은 상태를 의미함

* FLAG_MOBIL: 핸드폰 소유 여부
* work_phone: 업무용 전화 소유 여부
* phone: 전화 소유 여부
* email: 이메일 소유 여부
* occyp_type: 직업 유형
* family_size: 가족 규모
* begin_month: 신용카드 발급 월
							데이터 수집 당시 (0)부터 역으로 셈, 즉, -1은 데이터 수집일 한 달 전에 신용카드를 발급함을 의미

* credit: 사용자의 신용카드 대금 연체를 기준으로 한 신용도
							=> 낮을 수록 높은 신용의 신용카드 사용자를 의미함

#### 3. shape

In [10]:
# train 데이터의 행과 열의 개수를 확인합니다.
train.shape

(26457, 20)

## Stage 2. 데이터 확인

### Step 1. 데이터프레임 정보 확인

#### 1. info()

In [11]:
# train 데이터프레임의 정보를 확인합니다.
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26457 entries, 0 to 26456
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          26457 non-null  int64  
 1   gender         26457 non-null  object 
 2   car            26457 non-null  object 
 3   reality        26457 non-null  object 
 4   child_num      26457 non-null  int64  
 5   income_total   26457 non-null  float64
 6   income_type    26457 non-null  object 
 7   edu_type       26457 non-null  object 
 8   family_type    26457 non-null  object 
 9   house_type     26457 non-null  object 
 10  DAYS_BIRTH     26457 non-null  int64  
 11  DAYS_EMPLOYED  26457 non-null  int64  
 12  FLAG_MOBIL     26457 non-null  int64  
 13  work_phone     26457 non-null  int64  
 14  phone          26457 non-null  int64  
 15  email          26457 non-null  int64  
 16  occyp_type     18286 non-null  object 
 17  family_size    26457 non-null  float64
 18  begin_

* `occyp_type` 변수에서 결측지가 있음을 확인할 수 있다.
* 예측할 `credit`변수를 제외하고 총 11개의 수치형 변수와 8개의 카테고리 변수가 존재

#### 2. describe()

In [12]:
# train 데이터의 기초 통계량을 간략하게 확인합니다.
train.describe()

index     child_num  income_total    DAYS_BIRTH  DAYS_EMPLOYED  \
count  26457.000000  26457.000000  2.645700e+04  26457.000000   26457.000000   
mean   13228.000000      0.428658  1.873065e+05 -15958.053899   59068.750728   
std     7637.622372      0.747326  1.018784e+05   4201.589022  137475.427503   
min        0.000000      0.000000  2.700000e+04 -25152.000000  -15713.000000   
25%     6614.000000      0.000000  1.215000e+05 -19431.000000   -3153.000000   
50%    13228.000000      0.000000  1.575000e+05 -15547.000000   -1539.000000   
75%    19842.000000      1.000000  2.250000e+05 -12446.000000    -407.000000   
max    26456.000000     19.000000  1.575000e+06  -7705.000000  365243.000000   

       FLAG_MOBIL    work_phone         phone         email   family_size  \
count     26457.0  26457.000000  26457.000000  26457.000000  26457.000000   
mean          1.0      0.224742      0.294251      0.091280      2.196848   
std           0.0      0.417420      0.455714      0.288013      0.916717   
min           1.0      0.000000      0.000000      0.000000      1.000000   
25%           1.0      0.000000      0.000000      0.000000      2.000000   
50%           1.0      0.000000      0.000000      0.000000      2.000000   
75%           1.0      0.000000      1.000000      0.000000      3.000000   
max           1.0      1.000000      1.000000      1.000000     20.000000   

        begin_month        credit  
count  26457.000000  26457.000000  
mean     -26.123294      1.519560  
std       16.559550      0.702283  
min      -60.000000      0.000000  
25%      -39.000000      1.000000  
50%      -24.000000      2.000000  
75%      -12.000000      2.000000  
max        0.000000      2.000000

* DAYS_BIRTH: 출생일
        데이터 수집 당시 (0)부터 역으로 셈, 즉, -1은 데이터 수집일 하루 전에 태어났음을 의미
* DAYS_EMPLOYED: 업무 시작일
        데이터 수집 당시 (0)부터 역으로 셈, 즉, -1은 데이터 수집일 하루 전부터 일을 시작함을 의미.
        양수 값은 고용되지 않은 상태를 의미함
* begin_month: 신용카드 발급 월
        데이터 수집 당시 (0)부터 역으로 셈, 즉, -1은 데이터 수집일 한 달 전에 신용카드를 발급함을 의미

In [14]:
# 각 변수들을 날짜 단위가 아닌 연 단위(만 나이)로 바꿔보세요 (begin_month 제외)
train.loc[train["DAYS_EMPLOYED"]> 0, "DAYS_EMPLOYED"] = 0
train["DAYS_BIRTH"] = (train["DAYS_BIRTH"] * -1) / 365.0
train["DAYS_EMPLOYED"] = (train["DAYS_EMPLOYED"] * -1) / 365.0
train["begin_month"] = (train["begin_month"] * -1)

#테스트 데이터도 마찬가지로 같은 변형을 적용
test.loc[test["DAYS_EMPLOYED"] > 0,"DAYS_EMPLOYED"] = 0
test["DAYS_BIRTH"] = (test["DAYS_BIRTH"] * -1) / 365.0
test["DAYS_EMPLOYED"] = (test["DAYS_EMPLOYED"] * -1) / 365.0
test["begin_month"] = (test["begin_month"] * -1)

#### 3. value_counts()

In [15]:
# train 데이터의 카테고리 컬럼들을 추출합니다.
categories = train.select_dtypes(exclude='number')

# 추출한 변수들의 값들의 비율을 확인합니다.
for feature in categories:
  print(train[feature].value_counts(normalize=True))
  print('-'*50)
# value_counts(normalize=True)로 파라미터를 설정하면 각 값들의 비율을 확인할 수 있다!

gender
F    0.668897
M    0.331103
Name: proportion, dtype: float64
--------------------------------------------------
car
N    0.620252
Y    0.379748
Name: proportion, dtype: float64
--------------------------------------------------
reality
Y    0.673924
N    0.326076
Name: proportion, dtype: float64
--------------------------------------------------
income_type
Working                 0.515743
Commercial associate    0.234418
Pensioner               0.168160
State servant           0.081415
Student                 0.000265
Name: proportion, dtype: float64
--------------------------------------------------
edu_type
Secondary / secondary special    0.680160
Higher education                 0.270703
Incomplete higher                0.038553
Lower secondary                  0.009714
Academic degree                  0.000869
Name: proportion, dtype: float64
--------------------------------------------------
family_type
Married                 0.687757
Single / not married    0.132139
Civ

### Step 2. 결측치 처리하기

In [16]:
# train 데이터의 `occyp_type` 변수의 결측치들을 "No job"으로 채워주세요
train['occyp_type'] = train['occyp_type'].fillna('No job')

# 마찬가지로 테스트 데이터 또한 결측치 처리를 해주도록 하겠습니다!
test["occyp_type"] = test["occyp_type"].fillna("No job")

In [17]:
# occyp_type 변수의 값의 비율을 백분율로 확인
round(train["occyp_type"].value_counts(normalize=True) * 100, 2)

occyp_type
No job                   30.88
Laborers                 17.05
Core staff               10.00
Sales staff               9.60
Managers                  8.19
Drivers                   5.95
High skill tech staff     3.93
Accountants               3.41
Medicine staff            3.27
Cooking staff             1.73
Security staff            1.60
Cleaning staff            1.52
Private service staff     0.92
Low-skill Laborers        0.48
Waiters/barmen staff      0.47
Secretaries               0.37
Realty agents             0.24
HR staff                  0.23
IT staff                  0.15
Name: proportion, dtype: float64

### Step 3. 전처리 데이터 저장하기

In [ ]:
# 전처리된 train, test 데이터를 각각 저장해줍니다.

# train.to_csv("train.csv", index=False)
# test.to_csv("test.csv", index=False)

## Stage 3. 모델링 기초

1. Feature & Target

        첫 번째는 Feature와 Target입니다.
        feature는 예측을 위해 모델이 학습하는 변수들을 말합니다! 독립변수 X라고도 불립니다.
        target은 모델이 예측하고자 하는 대상입니다!
        모델은 feature를 통해 target과의 관계, 패턴 등을 학습하고 결과를 만들어냅니다. 종속변수 Y라고도 불립니다.
        
        train 데이터셋에서 Target은 credit 변수, Feature는 나머지 변수가 됩니다!

2. Problem Type (Regression vs Classification)
        두 번째는 예측 문제 유형의 정의입니다!
        예측 문제는 주로 회귀(Regression)과 분류(Classification) 문제로 나눌 수 있습니다.
        둘 중 어떤 문제로 정의할지는 Target 데이터의 특성에 따라 결정됩니다.

  * 회귀(Regression)
          - Target 데이터가 연속적인 값을 가질 때 이를 예측하기 위해 사용됩니다.
          - 대표적으로 주택 가격 예측 등의 문제가 있습니다.
  * 분류(Classification)
          - Target 데이터가 여러 개의 이산적인 값을 가질 때 이를 예측하기 위해 사용됩니다.
          - 대표적으로 스팸 메일 분류, 이미지 분류 등의 문제가 있습니다.

In [18]:
# credit 변수를 살펴봅니다
train["credit"].value_counts()

credit
2.0    16968
1.0     6267
0.0     3222
Name: count, dtype: int64

* Target이 0, 1, 2의 세 가지 결과로 나눠지기 때문에 분류(Classification) 알고리즘이 좋을 것 같네요!

## Stage 3. 모델링 기초
### Step 1. 모델링

In [20]:
# sklearn 패키지의 tree 모듈에서 DecisionTreeClassifier 객체를 불러옵니다
from sklearn.tree import DecisionTreeClassifier

# model이라는 변수에 DecisionTreeClassifier 객체를 할당합니다
model = DecisionTreeClassifier(random_state=42)

#### Step 2. 레이블 인코딩

* 레이블 인코딩이란?
      레이블 인코딩(Label Encoding)은 카테고리형 데이터를 수치형 데이터로 변환하는 방법 중 하나입니다.
      머신 러닝 알고리즘에 입력 데이터로 사용하기 위해, 문자열 형태의 카테고리를 숫자로 변환하는 프로세스를 의미합니다

In [21]:
# sklearn 패키지의 preprocessing 모듈에서 LabelEncoder 함수를 불러옵니다
from sklearn.preprocessing import LabelEncoder

# le라는 변수에 LabelEncoder 함수를 할당합니다
le = LabelEncoder()

In [23]:
# LabelEncoder로 object 타입 변수들을 수치형 데이터로 변환해줍니다
cols = train.select_dtypes(include=['object']).columns

for col in cols:
    train[col] = le.fit_transform(train[col])
    test[col] = le.fit_transform(test[col])

In [24]:
# 변환 결과를 확인합니다
train.head()

index  gender  car  reality  child_num  income_total  income_type  \
0      0       0    0        0          0      202500.0            0   
1      1       0    0        1          1      247500.0            0   
2      2       1    1        1          0      450000.0            4   
3      3       0    0        1          0      202500.0            0   
4      4       0    1        1          0      157500.0            2   

   edu_type  family_type  house_type  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  \
0         1            1           2   38.079452      12.901370           1   
1         4            0           1   31.178082       4.219178           1   
2         1            1           1   52.293151      12.147945           1   
3         4            1           1   41.336986       5.731507           1   
4         1            1           1   41.197260       5.767123           1   

   work_phone  phone  email  occyp_type  family_size  begin_month  credit  
0           0      0      0          12          2.0          6.0     1.0  
1           0      0      1           8          3.0          5.0     1.0  
2           0      1      0          10          2.0         22.0     2.0  
3           0      1      0          15          2.0         37.0     0.0  
4           0      0      0          10          2.0         26.0     2.0

### Step 3. 데이터 나누기

In [25]:
# train 데이터를 독립변수 X와 종속변수 y로 나눠줍니다
X = train.drop(columns="credit")
y = train["credit"]

In [27]:
# sklearn 패키지의 train_test_split 함수를 불러옵니다
from sklearn.model_selection import train_test_split

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, random_state=42)

### Step 4. 모델 학습 및 평가

In [28]:
# model의 학습을 진행합니다
model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [29]:
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_valid)
acc = accuracy_score(y_valid, y_pred)
print(f"Accuracy: {round(acc, 4)}")

Accuracy: 0.5865


In [30]:
# 분류 문제에서는 주로 평가를 위해 오차 행렬(confusion matrix)을 사용
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_valid)
cm = confusion_matrix(y_valid, y_pred, labels=[0, 1, 2])
print(cm)

[[ 164  156  330]
 [ 174  521  567]
 [ 379  582 2419]]


| |true_0|true_1|true_2|
|---|---|---|---|
|pred_0|164  |156  |330|
|pred_1|174  |521  |567|
|pred_2|379  |582 |2419|
| |717 |1259 |3316 |*이탤릭체 텍스트*

* credit을 0은 164/717 = 22.87%, 1은 521/1259 = 41.38%, 2는 2419/3316 = 72.95%의
비율만 제대로 예측한 것을 볼 수 있네요.

### Step 5. 예측하기 및 파일 제출

In [31]:
# 확률로 결과를 저장하기 위해 predict_proba() 메서드로 예측결과를 만들도록 할게요!
pred = model.predict_proba(test)

submission.iloc[:,1:] = pred
submission

index  0  1  2
0     26457  0  0  1
1     26458  0  0  1
2     26459  1  0  0
3     26460  0  0  1
4     26461  0  1  0
...     ... .. .. ..
9995  36452  0  0  1
9996  36453  1  0  0
9997  36454  0  1  0
9998  36455  0  1  0
9999  36456  1  0  0

[10000 rows x 4 columns]

In [ ]:
submission.to_csv("submission1.csv", index=False)